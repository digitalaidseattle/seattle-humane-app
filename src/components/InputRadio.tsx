import { RadioButton, RadioButtonProps } from "primereact/radiobutton";

interface InputRadioProps extends RadioButtonProps {
  /** The label for the element */
  label: string
  /** Whether this radio item is selected */
  checked: boolean
}

/**
 * 
 * @param props {@link InputRadioProps}
 * @returns A controlled radio element
 */
export default function InputRadio(props: InputRadioProps) {
  const { id, name, value, onChange, disabled, label, checked } = props
  return (
    <div className="flex align-items-center">
      <RadioButton
        inputId={id}
        name={name}
        value={value}
        disabled={disabled}
        onChange={(...args) => !disabled && onChange(...args)}
        checked={checked}
      />
      <label htmlFor={id} className="ml-2">
        {label}
      </label>
    </div>
  )
}
