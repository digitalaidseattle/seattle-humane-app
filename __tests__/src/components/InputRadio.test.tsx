import '@testing-library/jest-dom';
import { render, screen, fireEvent, act } from '@testing-library/react';
import InputRadio from "@components/InputRadio";
import { useEffect, useState } from 'react';
import { RadioButton } from 'primereact/radiobutton';

describe('InputRadio', () => {
  const fakeId = 'fake-id'
  const labelText = 'fake-label'
  const fakeValue = 'fake-value'

  it.each([['unchecked', false], ['checked', true]])('should render a %s radio button when the checked prop is %s', (label, defaultChecked) => {
    //* Arrange
    render(<InputRadio
      id={fakeId}
      checked={defaultChecked}
      label={labelText}
      value={fakeValue}
    />)
    const labelElement = screen.getByText(labelText)
    const radioButton = screen.getByLabelText(labelText)
    const component = radioButton.parentElement
    //* Assert
    expect(labelElement).toBeInTheDocument()
    expect(radioButton).toBeInTheDocument()

    if (defaultChecked) expect(radioButton).toBeChecked()
    else expect(radioButton).not.toBeChecked()

    // Snapshot added for friction when changing markup/classes/styles
    expect(component).toMatchSnapshot()
  })

  it('should call onChange with the target element and value when the radio is checked', () => {
    //* Arrange
    const onChangeProp = jest.fn()
    render(<InputRadio
      id={fakeId}
      checked={false}
      label={labelText}
      value={fakeValue}
      onChange={onChangeProp}
    />)
    //* Act
    const radioButton = screen.getByLabelText(labelText)
    fireEvent.click(radioButton)
    //* Assert
    expect(onChangeProp).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ checked: true, value: fakeValue })
      })
    )
  })

  it('should not call onChange when disabled', () => {
    //* Arrange
    const onChangeProp = jest.fn()
    render(<InputRadio
      id={fakeId}
      checked={false}
      value={fakeValue}
      disabled={true}
      label={labelText}
      onChange={onChangeProp}
    />)
    //* Act
    const radioButton = screen.getByLabelText(labelText)
    fireEvent.click(radioButton)
    //* Assert
    expect(onChangeProp).not.toHaveBeenCalled()
  })

  it('should update as expected when consumed by a stateful component', () => {
    //* Arrange
    function MockComponent() {
      const [value, setValue] = useState('val1')
      return <>
        <InputRadio
          id={fakeId + '_1'}
          checked={value === 'val1'}
          label={labelText + '_1'}
          value={'val1'}
          onChange={(e) => setValue(e.value)}
        />
        <InputRadio
          id={fakeId + '_2'}
          checked={value === 'val2'}
          label={labelText + '_2'}
          value={'val2'}
          onChange={(e) => setValue(e.value)}
        />
      </>
    }
    render(<MockComponent />)
    const radioButton1 = screen.getByLabelText(labelText + '_1')
    expect(radioButton1).toHaveAttribute('checked')
    const radioButton2 = screen.getByLabelText(labelText + '_2')
    //* Act
    fireEvent.click(radioButton2)
    //* Assert
    expect(radioButton1).not.toHaveAttribute('checked')
    expect(radioButton2).toHaveAttribute('checked')
  })
})