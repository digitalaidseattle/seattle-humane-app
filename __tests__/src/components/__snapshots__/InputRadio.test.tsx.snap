// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRadio should render a checked radio button when the checked prop is true 1`] = `
<div
  class="p-hidden-accessible"
  data-pc-section="hiddeninputwrapper"
>
  <input
    checked=""
    data-pc-section="hiddeninput"
    id="fake-id"
    type="radio"
  />
</div>
`;

exports[`InputRadio should render a unchecked radio button when the checked prop is false 1`] = `
<div
  class="p-hidden-accessible"
  data-pc-section="hiddeninputwrapper"
>
  <input
    data-pc-section="hiddeninput"
    id="fake-id"
    type="radio"
  />
</div>
`;
